File: supabase/functions/ks2_operations/vat_calc.ts
/** Параметры НДС договора. */
export type ContractVatTerms = {
  vatRate: number;
  isVatIncluded: boolean;
};

/** Округление до копеек. */
export function roundMoney(value: number): number {
  if (!Number.isFinite(value)) return 0;
  return Math.round(value * 100) / 100;
}

/** НДС от суммы по правилам договора. */
export function computeVatAmount(
  baseAmount: number,
  terms: ContractVatTerms,
): number {
  const { vatRate, isVatIncluded } = terms;
  if (vatRate <= 0 || baseAmount <= 0) return 0;
  const raw = isVatIncluded
    ? (baseAmount * vatRate) / (100 + vatRate)
    : (baseAmount * vatRate) / 100;
  return roundMoney(raw);
}

/** Итог строк ВОР без НДС → `amount` + начисленный сверху `vat_amount`. */
export function splitActAmountForStorage(
  lineTotal: number,
  terms: ContractVatTerms,
): { amount: number; vatAmount: number } {
  if (!Number.isFinite(lineTotal) || lineTotal <= 0) {
    return { amount: 0, vatAmount: 0 };
  }
  const amount = roundMoney(lineTotal);
  if (terms.vatRate <= 0) {
    return { amount, vatAmount: 0 };
  }
  return {
    amount,
    vatAmount: computeVatAmount(amount, {
      vatRate: terms.vatRate,
      isVatIncluded: false,
    }),
  };
}

type VatTermsRow = {
  vat_rate: string | number | null;
  is_vat_included: boolean | null;
};

/** Загружает ставку и режим НДС договора. */
export async function loadContractVatTerms(
  supabase: {
    from: (table: string) => {
      select: (cols: string) => {
        eq: (col: string, val: string) => {
          eq: (col: string, val: string) => {
            maybeSingle: () => Promise<{
              data: VatTermsRow | null;
              error: unknown;
            }>;
          };
        };
      };
    };
  },
  contractId: string,
  companyId: string,
): Promise<ContractVatTerms> {
  const { data, error } = await supabase
    .from("contracts")
    .select("vat_rate, is_vat_included")
    .eq("id", contractId)
    .eq("company_id", companyId)
    .maybeSingle();

  if (error) throw error;
  if (!data) throw new Error("Договор не найден");

  const rateRaw = data.vat_rate;
  const vatRate = rateRaw == null ? 0 : Number(rateRaw);
  return {
    vatRate: Number.isFinite(vatRate) ? vatRate : 0,
    isVatIncluded: data.is_vat_included ?? true,
  };
}

File: lib/domain/utils/vat_calc.dart
/// Расчёт НДС по правилам договора (`vat_rate`, `is_vat_included`).
///
/// Используется в карточке договора, актах КС-2 и сводках.
library;

/// Параметры НДС договора для расчёта актов.
class ContractVatTerms {
  /// Создаёт параметры НДС.
  const ContractVatTerms({required this.vatRate, required this.isVatIncluded});

  /// Ставка НДС, %.
  final double vatRate;

  /// `true` — НДС включён в сумму договора; `false` — начисляется сверху.
  ///
  /// Для актов КС-2 не используется: сумма строк ВОР всегда без НДС.
  final bool isVatIncluded;
}

/// Сумма акта и НДС для записи в `contract_acts`.
class ActAmountVatSplit {
  /// Создаёт пару сумм для сохранения акта.
  const ActAmountVatSplit({required this.amount, required this.vatAmount});

  /// База акта без начисляемого «сверху» НДС (при включённом НДС — нетто).
  final double amount;

  /// Сумма НДС (для `total_to_pay` складывается с [amount]).
  final double vatAmount;
}

/// Округление денежной суммы до копеек.
double roundMoney(double value) {
  if (value.isNaN || value.isInfinite) return 0;
  return (value * 100).round() / 100;
}

/// НДС от суммы [baseAmount] по ставке договора.
///
/// При [isVatIncluded] `baseAmount` трактуется как сумма с учётом НДС
/// (извлекается доля НДС). Иначе НДС начисляется сверху.
double computeVatAmount({
  required double baseAmount,
  required double vatRate,
  required bool isVatIncluded,
}) {
  if (vatRate <= 0 || baseAmount <= 0) return 0;
  final raw = isVatIncluded
      ? baseAmount * vatRate / (100 + vatRate)
      : baseAmount * vatRate / 100;
  return roundMoney(raw);
}

/// Разбивает итог строк акта КС-2 на поля `amount` и `vat_amount`.
///
/// [lineTotal] — сумма строк ВОР/КС-2 **без НДС**. НДС начисляется сверху
/// по [ContractVatTerms.vatRate] (режим «включён в договор» на карточку договора
/// к строкам акта не распространяется).
ActAmountVatSplit splitActAmountForStorage({
  required double lineTotal,
  required ContractVatTerms vatTerms,
}) {
  if (lineTotal.isNaN || lineTotal <= 0) {
    return const ActAmountVatSplit(amount: 0, vatAmount: 0);
  }
  final amount = roundMoney(lineTotal);
  if (vatTerms.vatRate <= 0) {
    return ActAmountVatSplit(amount: amount, vatAmount: 0);
  }
  final vatAmount = computeVatAmount(
    baseAmount: amount,
    vatRate: vatTerms.vatRate,
    isVatIncluded: false,
  );
  return ActAmountVatSplit(amount: amount, vatAmount: vatAmount);
}
